src/Cli/dotnet/commands/dotnet-publish/Program.cs

```diff
@@ -43,30 +64,64 @@ public static PublishCommand FromParseResult(ParseResult parseResult, string msb
-            msbuildArgs.AddRange(slnOrProjectArgs ?? Array.Empty<string>());
-
             bool noRestore = parseResult.HasOption(PublishCommandParser.NoRestoreOption)
                           || parseResult.HasOption(PublishCommandParser.NoBuildOption);
 
+
+            var publishProfileProperties = DiscoverPropertiesFromPublishProfile(parseResult);
+            var standardMSbuildProperties = CreatePropertyListForPublishInvocation(parseResult);
             return new PublishCommand(
-                msbuildArgs,
+                // properties defined by the selected publish profile should override any other properties,
+                // so they should be added after the other properties.
+                standardMSbuildProperties.Concat(publishProfileProperties),
                 noRestore,
-                msbuildPath);
+                msbuildPath
+            );
+
+            List<string> CreatePropertyListForPublishInvocation(ParseResult parseResult) {
+                var msbuildArgs = new List<string>()
+                    {
+                        "-target:Publish",
+                        "--property:_IsPublishing=true" // This property will not hold true for MSBuild /t:Publish. VS should also inject this property when publishing in the future.
+                    };
+
+                IEnumerable<string> slnOrProjectArgs = parseResult.GetValueForArgument(PublishCommandParser.SlnOrProjectArgument);
+
+                CommonOptions.ValidateSelfContainedOptions(parseResult.HasOption(PublishCommandParser.SelfContainedOption),
+                    parseResult.HasOption(PublishCommandParser.NoSelfContainedOption));
+
+                msbuildArgs.AddRange(parseResult.OptionValuesToBeForwarded(PublishCommandParser.GetCommand()));
+                ReleasePropertyProjectLocator projectLocator = new ReleasePropertyProjectLocator(Environment.GetEnvironmentVariable(EnvironmentVariableNames.ENABLE_PUBLISH_RELEASE_FOR_SOLUTIONS) != null);
+                msbuildArgs.AddRange(projectLocator.GetCustomDefaultConfigurationValueIfSpecified(parseResult, MSBuildPropertyNames.PUBLISH_RELEASE, slnOrProjectArgs, PublishCommandParser.ConfigurationOption) ?? Array.Empty<string>());
+                msbuildArgs.AddRange(slnOrProjectArgs ?? Array.Empty<string>());
+
+                return msbuildArgs;
+            }
+
+            List<string> DiscoverPropertiesFromPublishProfile(ParseResult parseResult)
+            {
+                ReleasePropertyProjectLocator projectLocator = new ReleasePropertyProjectLocator(Environment.GetEnvironmentVariable(EnvironmentVariableNames.ENABLE_PUBLISH_RELEASE_FOR_SOLUTIONS) != null);
+                var cliProps = projectLocator.GetGlobalPropertiesFromUserArgs(parseResult);
+                var projectInstance = projectLocator.GetTargetedProject(parseResult.GetValueForArgument(PublishCommandParser.SlnOrProjectArgument), cliProps, MSBuildPropertyNames.PUBLISH_RELEASE);
```

Comment:
Note that this will be another +~18-25% performace regression in time to publish which I don't think we want. If we were to do this, I think it would be better to create something like a singleton wrapper which only ever instantiates or pre-evaluates the project once.